### DISCUSSION TITLE
LPS + Flow deck estimated z position

### DISCUSSION THREAD

**Discussion by yuhsiangsu** [bitcraze/discussions#638] [Q&A]
Hi,

I am using LPS and Flow deck together to fly the Crazyflies. However, I am a bit confused about the estimated position in the z-direction. 

If I understand correctly, when we only use the Flow deck, it detects the relative z distance between the Crazyflie and the underneath surface. This means we will get relative z-direction data between the Crazyflie and the underneath surface from the estimate.z. 

However, if we use both LPS and Flow deck. What exactly do we get from the estimate.z? Is it the relative z-direction data between the Crazyflie and the underneath surface? Or is it the absolute global z position of the Crazyflie with respect to the origin? 

In my experiment, I set the ground as the origin of the global z direction, and I wish to get the absolute z position of the Crazyflie with respect to the ground when using both LPS and the Flow deck. I managed to get the absolute x and y positions of the Crazyflie with respect to the origin from the estimate.x and estimate.y, but it seems like I only got the relative z-direction data between the Crazyflie and the underneath surface from the estimate.z. That is when an object exists between the Crazyflie and the ground, the estimate.z returns the relative z-direction data between the Crazyflie and the object but not the absolute z position of the Crazyflie with respect to the ground. 

Please help me with this doubt. Thank you.

**Comment by krichardsson** [comment]
Hi!

All sensor data is pushed into the kalman estimator to form a full estimate of the system state. The standard deviation for each measurement is pushed together with the values and is used by the estimator to decide how much to "trust" the measurement, and thus how much it will effect the current estimate. 
In your case, the distance and optical flow sensors on the Flow deck are very accurate and the standard deviation low, while the Loco system data is fairly noisy and has a higher standard deviation. This means that the estimator will be influenced more by the flow deck sensors compared to the Loco system. 
In the current implementation, it is pretty much assumed that there is a flat floor under the Crazyflie and the distance from the ranging sensor is treated as absolute Z. 

> If I understand correctly, when we only use the Flow deck, it detects the relative z distance between the Crazyflie and the underneath surface.

The [`estimate.z`](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/logs/#kalman) is actually the estimated global Z. If you use the Flow deck and fly in a horizontal, straight line over a rising surface, the estimated absolute Z will decrease. 

It sounds like the surface under the Crazyflie in your case is not a flat floor? This is a bit problematic as the distance sensor will have a much stronger influence on the z-estimate than the Loco system.

**Reply by yuhsiangsu** [reply]
Hi Kristoffer,

Thank you for your answer.

If this is the case, would it be helpful to remove the Flow deck? So the distance sensor won't affect the Kalman filter.

**Reply by krichardsson** [reply]
I think it depends on what your needs are. Could you tell us a bit about your use case?

**Reply by yuhsiangsu** [reply]
Hi Kristoffer,

Sorry for the late reply.

I have two different cases:

1. I would like to fly the Crazyflie over an obstacle underneath while keeping a constant altitude. Similar to the case in this [blog](https://www.bitcraze.io/tag/flow-deck/).
2. I want the Crazyflie to carry a cable-suspended payload underneath, and I want the Crayflie to fly at a constant altitude with respect to the floor. 


Thanks.

**Reply by krichardsson** [reply]
> I would like to fly the Crazyflie over an obstacle underneath while keeping a constant altitude. Similar to the case in this [blog](https://www.bitcraze.io/tag/flow-deck/).

This will unfortunately not work with the default implementation + flow deck. Either remove the deck (easy) or update the kalman filter (complex) 

> I want the Crazyflie to carry a cable-suspended payload underneath, and I want the Crayflie to fly at a constant altitude with respect to the floor.

This should work out of the box, from a positioning point of view. Note that the load will change the behavior of the Crazyflie and you might need to modify the controller for good performance.
